From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Thu, 7 Dec 2017 02:12:47 +0100
Subject: workspace: Disconnect from window signals on destruction

Avoid to try to destroy the window clone content more than once
when a window is destroyed, and do it in proper order.

(cherry picked from commit 6da7b3bfdede17a9e3d7f58c7a32e11f3885c132)
Bug: https://bugzilla.gnome.org/show_bug.cgi?id=791233
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/bionic/+source/gnome-shell/+bug/1747566
Forwarded: https://gitlab.gnome.org/GNOME/gnome-shell/merge_requests/4
---
 js/ui/workspace.js | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/js/ui/workspace.js b/js/ui/workspace.js
index 8b56932..7d6339b 100644
--- a/js/ui/workspace.js
+++ b/js/ui/workspace.js
@@ -139,14 +139,8 @@ var WindowClone = new Lang.Class({
 
         this._windowClone._updateId = this.metaWindow.connect('size-changed',
             this._onRealWindowSizeChanged.bind(this));
-        this._windowClone._destroyId =
-            this.realWindow.connect('destroy', () => {
-                // First destroy the clone and then destroy everything
-                // This will ensure that we never see it in the
-                // _disconnectSignals loop
-                this._windowClone.destroy();
-                this.destroy();
-            });
+        this._windowClone._destroyId = this.realWindow.connect('destroy',
+            this.destroy.bind(this));
 
         this._updateAttachedDialogs();
         this._computeBoundingBox();
@@ -310,6 +304,12 @@ var WindowClone = new Lang.Class({
     },
 
     destroy() {
+        // First destroy the clone and then destroy everything
+        // This will ensure that we never see it in the _disconnectSignals loop
+        this.metaWindow.disconnect(this._windowClone._updateId);
+        this.realWindow.disconnect(this._windowClone._destroyId);
+        this._windowClone.destroy();
+
         this.actor.destroy();
     },
 
